### PR TITLE
Check whether locals are too large instead of whether accesses into them are too large

### PR DESCRIPTION
**Pull request by oli-obk**
Essentially this stops const prop from attempting to optimize

```rust
let mut x = [0_u8; 5000];
x[42] = 3;
```

I don't expect this to be a perf improvement without #73656 (which is also where the lack of this PR will be a perf regression).

r? @wesleywiser

### CODE REVIEW

**Comment by oli-obk** [comment]
@bors try @rust-timer queue

but even if I don't expect any perf changes, let's make sure

**Review by wesleywiser — APPROVED** [review]
LGTM! Can we add a mir-opt test showing that 

```rust
let mut x = [0_u8; 5000];
x[42] = 3;
```

does not propagate? r=me with a test.

**Comment by oli-obk** [comment]
I had to update the output of a clippy lint. clippy builds with `-Zmir-opt-level=0`, which throws off const prop completely.

@bors r=wesleywiser

**Comment by oli-obk** [comment]
I made some changes to the mir-opt test suite. It now prints the name of the mir dump file it's expecting and the path to the file which requested the mir dump

File: src/tools/compiletest/src/runtest.rs [hunk 1]
```diff
@@ -3239,8 +3239,19 @@ impl<'test> TestCx<'test> {
     }
 
     fn diff_mir_files(&self, before: PathBuf, after: PathBuf) -> String {
-        let before = self.get_mir_dump_dir().join(before);
-        let after = self.get_mir_dump_dir().join(after);
+        let to_full_path = |path: PathBuf| {
+            let full = self.get_mir_dump_dir().join(&path);
+            if !full.exists() {
+                panic!(
```
Comment (xldenis) [inline src/tools/compiletest/src/runtest.rs:3245]: 👍 there are a couple other places in `check_mir_dump` that also produce confusing / less useful errors but that could be fixed in a seperate pr.

**Review by wesleywiser — APPROVED** [review]

**Pull request merged**